File: 프로그래머스/level2/dfs_bfs/단어변환.py
def isOneLetterDiff(a, b):
    count = 0 
    for i in range(len(a)):
        if a[i] != b[i]:
            count += 1
            
    return count == 1

def solution(begin, target, words):
    answer = 100
    visited = []
    
    def dfs(crnt, count):
        nonlocal answer
        
        if crnt == target:
            answer = min(answer, count)
            return
        
        for word in words:
            if word not in visited and isOneLetterDiff(crnt, word):
                visited.append(word)
                dfs(word, count+1)
                visited.pop()
        
    dfs(begin, 0)
    if answer == 100:
        return 0
    return answer

File: 프로그래머스/dfs_bfs/단어변환.js
function isOneLetterDiff(from, to) {
  let count = 0;
  for (let i in from) {
    if (from[i] !== to[i]) ++count;
  }
  return count === 1;
}

function solution(begin, target, words) {
  var answer = 50;

  function dfs(cword, count, visited) {
    if (cword === target) {
      answer = Math.min(count, answer);
      return;
    }

    visited.add(cword);
    for (let w of words) {
      if (isOneLetterDiff(cword, w) && !visited.has(w)) {
        dfs(w, count + 1, visited);
        visited.delete(w);
      }
    }
  }

  dfs(begin, 0, new Set());

  return answer < 50 ? answer : 0;
}

console.log(solution("hit", "cog", ["hot", "dot", "dog", "lot", "log", "cog"]));
console.log(solution("hit", "cog", ["hot", "dot", "dog", "lot", "log"]));
